keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,942 @@
+2. Filter: 
+    * For each node being evaluated, the plugin will retrieve the cached requirements and pass them along with the Node object to a matching function in the shared library.
+    * The library itself is responsible for handling all complex logic, including the node's opt-in check. The plugin simply acts on the boolean result returned by the library to either filter the node or pass it.
+
+#### Performance Considerations
+
+Introducing this plugin adds a new step to the scheduling cycle, which may have an impact on scheduling throughput. However, this trade-off is acceptable because it prevents the greater inefficiency of scheduling a pod onto a node where it cannot run.
+
+
+### Admission Controller Changes
+
+To enable the validation of API requests against Node Capabilities, this KEP proposes the introduction of a new admission controller plugin `NodeCapabilityValidator` that will be enabled when the `NodeCapabilities` feature gate is active. This admission controller cross-references the objects, i.e., looking  up the `Node` object that a `Pod` is bound to (spec.nodeName) and runs the validation checks.
+
+The admission controller workflow will be as follows:
+*   Inspect an incoming pod update request and verifies that the pod is already bound to a node by checking that `pod.spec.nodeName` is set. If not, it takes no action.
+*   Call the shared library to infer the capability requirements.
+*   Retrieves the Node object corresponding to `pod.spec.nodeName` and pass it along with the inferred requirements  to the shared library's matching function.
+*   If the function returns false, the admission controller rejects the request.
+
+### Capability Lifecycle and Deprecation
+
+To prevent the `node.status.capabilities` map from becoming an ever growing collection of features,  every capability tied to a feature's lifecycle must be automatically removed from the `node.status` object.
+
+The lifecycle stages are as follows:
+
+* Introduction (Alpha/Beta): A new capability that is introduced alongside a new feature. Kubelet begins reporting the capability on nodes where the feature is enabled and active. Control plane components, via the shared library, start using this capability to make decisions.
+* Graduation (GA): When the feature graduates to GA, the Kubelet continues to report the capability. This is necessary to manage version skew, allowing the control plane to correctly identify older nodes that do not yet have the GA feature.
+* Automated Deprecation (Post-GA): Kubelet automatically stop reporting the capability after the feature has been GA for a duration that exceeds the cluster's supported version skew. The capability check is bypassed in the shared library based on consumer component (e.g., kube-scheduler) version and feature gate graduation version.
```

Comment:
This requires clarification. Maybe with the specific versions on how to calculate supported version skews. Does this statement suggests that the capability will be removed only after GA + 3 versions? And after this, the logic is removed from both - control plane and kubelet at the same time?